src/states/establishing_node.rs

```diff
@@ -0,0 +1,463 @@
+// Copyright 2019 MaidSafe.net limited.
+//
+// This SAFE Network Software is licensed to you under The General Public License (GPL), version 3.
+// Unless required by applicable law or agreed to in writing, the SAFE Network Software distributed
+// under the GPL Licence is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied. Please review the Licences for the specific language governing
+// permissions and limitations relating to use of the SAFE Network Software.
+
+use super::{
+    common::{
+        proxied, Approved, Base, Bootstrapped, NotEstablished, Relocated, RelocatedNotEstablished,
+    },
+    node::Node,
+};
+use crate::{
+    ack_manager::AckManager,
+    cache::Cache,
+    chain::{Chain, GenesisPfxInfo, ProvingSection, SectionInfo},
+    error::RoutingError,
+    id::{FullId, PublicId},
+    messages::{DirectMessage, HopMessage, Message, RoutingMessage},
+    outbox::EventBox,
+    parsec::{self, Block, GossipMessage, Parsec},
+    peer_manager::{Peer, PeerManager, PeerState},
+    routing_message_filter::RoutingMessageFilter,
+    routing_table::{Authority, Prefix},
+    state_machine::{State, Transition},
+    time::{Duration, Instant},
+    timer::Timer,
+    xor_name::XorName,
+    Service,
+};
+use itertools::Itertools;
+use std::{
+    collections::BTreeSet,
+    fmt::{self, Display, Formatter},
+};
+
+const POKE_TIMEOUT: Duration = Duration::from_secs(60);
+
+pub struct EstablishingNode {
+    ack_mgr: AckManager,
+    cache: Box<Cache>,
+    chain: Chain,
+    crust_service: Service,
+    full_id: FullId,
+    gen_pfx_info: GenesisPfxInfo,
+    /// Routing messages addressed to us that we cannot handle until we are established.
+    msg_backlog: Vec<RoutingMessage>,
+    notified_nodes: BTreeSet<PublicId>,
+    parsec: Parsec,
+    peer_mgr: PeerManager,
+    poke_timer_token: u64,
+    routing_msg_filter: RoutingMessageFilter,
+    timer: Timer,
+}
+
+impl EstablishingNode {
+    #[allow(clippy::too_many_arguments)]
+    pub fn from_proving_node(
+        ack_mgr: AckManager,
+        cache: Box<Cache>,
+        crust_service: Service,
+        full_id: FullId,
+        gen_pfx_info: GenesisPfxInfo,
+        min_section_size: usize,
+        msg_backlog: Vec<RoutingMessage>,
+        notified_nodes: BTreeSet<PublicId>,
+        peer_mgr: PeerManager,
+        routing_msg_filter: RoutingMessageFilter,
+        timer: Timer,
+    ) -> Self {
+        let public_id = *full_id.public_id();
+        let poke_timer_token = timer.schedule(POKE_TIMEOUT);
+
+        let parsec = parsec::create(full_id.clone(), &gen_pfx_info);
+        let chain = Chain::new(min_section_size, public_id, gen_pfx_info.clone());
+
+        let node = Self {
+            ack_mgr,
+            cache,
+            chain,
+            crust_service,
+            full_id,
+            gen_pfx_info,
+            msg_backlog,
```

Comment:
`routing_msg_backlog: vec![]`

This was initialized with empty vector, while the msg_queue field took the initial content.
Now, we initialize `routing_msg_backlog` with what would have gone to `msg_queue`.

I'm not clear why this is correct?
Do we have test for this, what is that about?

Reply:
`msg_backlog` is for messages that are to be processed later, when we become `Node`. `msg_queue` is for messages that are to be processed now (-ish). Only `ProvingNode` and `EstablishingNode` need `msg_backlog` but don't need `msg_queue` (they either backlog an incoming message, or handle it immediately). `Node`, on the other hand, doesn't need `msg_backlog` (it handles all messages), but needs `msg_queue` (it needs to delay message handling. I think to prevent too deep/infinite recursion. Not 100% sure, but it was there before this PR and I didn't want to change it).

Reply:
I'm ok with what was there before (even if i don't necessarily understand all of it).

My question was specifically about what I saw as a change of behaviour: before Node started with empty msg_backlog, now the state that replace it EstablishingNode use what Node would take to populate msg_queue, and use it to populate msg_backlog.

Reply:
Turns out that if `ProvingNode` doesn't handle a message, `EstablishingNode` doesn't handle it either. So it's OK to just transfer `msg_backlog` from `ProvingNode` to `EstablishingNode` without giving `EstablishingNode` chance to handle them first.

Maybe this is something to be documented somewhere? Or alternatively, should we loop through `msg_backlog` when transitioning to `EstablishingNode` to make the code more future proof?

Reply:
I'd say if we probably want to keep the pass through behaviour in this PR:
I'd add the msg_queue field to EstablishingNode that is only in intialized in from_... and used in into_...

Then it is more clearly as it was before.

Reply:
But I fail to see how that would be functionally different from what is in the PR now. It would only make things more complex and confusing.

What I would consider doing is something like this: In `EstablishingNode::from_proving_node`, I would set `self.msg_backlog` to empty `Vec` initially, and then do:

```rust
for msg in msg_backlog {
    self.dispatch_routing_msg(msg, outbox);
}
```
That way, any message in `msg_backlog` that `ProvingNode` put there and that `EstablishingNode` actually knows how to handle, would be handled. The rest would be moved over to `self.msg_backlog`. Currently there are no such messages that `EstablishingNode` handles but `ProvingNode` does not, but it would make the code more robust in terms of future changes. I'm not convinced it's needed though.

Reply:
> It would only make things more complex and confusing.

Reply:
> It would only make things more complex and confusing.

Yes, I think I understand things better now.

`for msg in msg_backlog {` yes, i think it may be sensible. That seem to be more like the behaviour we had before this PR.

Also maybe change the name of the parameter to msg_queue.